src/meta/src/barrier/mod.rs

```diff
@@ -255,79 +330,122 @@ where
-                }
-                Err(e) => {
-                    notifiers
-                        .into_iter()
-                        .for_each(|notifier| notifier.notify_collection_failed(e.clone()));
-                    if self.enable_recovery {
-                        // If failed, enter recovery mode.
-                        let (new_epoch, actors_to_finish, finished_create_mviews) =
-                            self.recovery(state.prev_epoch, Some(command)).await;
-                        unfinished = UnfinishedNotifiers::default();
-                        unfinished.add(new_epoch.0, actors_to_finish, vec![]);
-                        for finished in finished_create_mviews {
-                            unfinished.finish_actors(finished.epoch, once(finished.actor_id));
+                    Err(e) => {
+                        notifiers
+                            .into_iter()
+                            .for_each(|notifier| notifier.notify_collection_failed(e.clone()));
+                        let mut is_over = false;
+                        if heap.read().await.is_empty() {
+                            is_over = true;
                         }
-
-                        state.prev_epoch = new_epoch;
-                    } else {
-                        panic!("failed to execute barrier: {:?}", e);
+                        barrier_send_tx_clone
+                            .send(BarrierSendResult::Err(command, e, is_over))
+                            .unwrap();
                     }
                 }
-            }
-
-            state.update(self.env.meta_store()).await.unwrap();
+                heap.write().await.pop();
+            });
         }
     }
 
     /// Running a scheduled command.
     async fn run_inner<'a>(
-        &self,
+        env: Arc<RwLock<MetaSrvEnv<S>>>,
+        metrics: Arc<MetaMetrics>,
         command_context: &CommandContext<'a, S>,
+        heap: Arc<RwLock<BinaryHeap<Reverse<u64>>>>,
+        hummock_manager: HummockManagerRef<S>,
+        wait_recovery: Arc<RwLock<bool>>,
     ) -> Result<Vec<InjectBarrierResponse>> {
-        let timer = self.metrics.barrier_latency.start_timer();
+        let timer = metrics.barrier_latency.start_timer();
+        if command_context.prev_epoch.0 != INVALID_EPOCH {
+            heap.write()
+                .await
+                .push(Reverse(command_context.prev_epoch.0));
+        }
 
         // Wait for all barriers collected
-        let result = self.inject_barrier(command_context).await;
+        let result = GlobalBarrierManager::inject_barrier(env, command_context).await;
         // Commit this epoch to Hummock
         if command_context.prev_epoch.0 != INVALID_EPOCH {
+            loop {
+                if let Some(Reverse(min_epoch)) = heap.read().await.peek() {
+                    if *min_epoch == command_context.prev_epoch.0 {
+                        break;
+                    }
+                }
+            }
```

Comment:
Busy loop checking will burn CPU for no benefits. Please use notify instead.

Reply:
yes ,I will modify it